main/files.go

```diff
@@ -29,34 +31,92 @@ func downloadAndProcessURL(ctx *log.Context, url, downloadDir string, cfg *handl
 		return "", fmt.Errorf(url + " is not a valid url") // url does not contain SAS to se can log it
 	}
 
-	downloadedFilePath := filepath.Join(downloadDir, fileName)
+	targetFilePath := filepath.Join(downloadDir, fileName)
 	scriptSAS := cfg.scriptSAS()
+
+	var scriptSASDownloadErr error = nil
+	var downloadedFilePath string = ""
 	if scriptSAS != "" {
-		downloadedFilePath, err = download.GetSASBlob(url, scriptSAS, downloadDir)
-	} else {
-		downloaders, err := getDownloaders(url)
-		if err == nil {
+		if UseMockSASDownloadFailure {
+			scriptSASDownloadErr = errors.New("Downloading script using SAS token failed.")
+		} else {
+			downloadedFilePath, scriptSASDownloadErr = download.GetSASBlob(url, scriptSAS, downloadDir)
+		}
+		// Download was successful using SAS. So use downloadedFilePath
+		if scriptSASDownloadErr == nil && downloadedFilePath != "" {
+			targetFilePath = downloadedFilePath
+		}
+	}
+
+	//If there was an error downloading using SAS URI or SAS was not provided, download using managedIdentity or publicly.
+	if scriptSASDownloadErr != nil || scriptSAS == "" {
+		downloaders, getDownloadersError := getDownloaders(url, cfg.SourceManagedIdentity)
+		if getDownloadersError == nil {
 			const mode = 0500 // we assume users download scripts to execute
-			_, err = download.SaveTo(ctx, downloaders, downloadedFilePath, mode)
+			_, err = download.SaveTo(ctx, downloaders, targetFilePath, mode)
+		} else {
+			return "", getDownloadersError
 		}
 	}
+
 	if err != nil {
 		return "", err
 	}
 
-	err = postProcessFile(downloadedFilePath)
+	err = postProcessFile(targetFilePath)
 	if err != nil {
 		return "", errors.Wrapf(err, "failed to post-process '%s'", fileName)
 	}
 
-	return downloadedFilePath, nil
+	return targetFilePath, nil
 }
 
-// getDownloader returns a downloader for the given URL based on whether the
-// storage credentials are empty or not.
-func getDownloaders(fileURL string) (
-	[]download.Downloader, error) {
-	return []download.Downloader{download.NewURLDownload(fileURL)}, nil
+// getDownloaders returns one or two downloaders (two if it is an Azure storage blob):
+// 1. Downloader for script using public URI.
+// 2. Downloader for script using managed identity.
+func getDownloaders(fileURL string, managedIdentity *RunCommandManagedIdentity) ([]download.Downloader, error) {
```

Comment:
files_test.go/Test_getDownloader_externalUrl is failing on line 34 because 2 downloaders are being returned. I'm not clear on the purpose of that test, but if this behavior is expected then that test should be fixed

Reply:
fixed